### ISSUE TITLE
Error in capture demo

### ISSUE THREAD

**Issue by jywarren** [publiclab/spectral-workbench.js#95]
Seeing this error on the capture demo at: https://publiclab.github.io/spectral-workbench.js/examples/capture/

```
Uncaught Error: Invalid dimensions for plot, width = 0, height = 200
    at getCanvasDimensions (jquery.flot.js:803)
    at setupCanvases (jquery.flot.js:853)
    at new Plot (jquery.flot.js:224)
    at Function.$.plot (jquery.flot.js:2690)
    at getRow (capture.js:341)
```

cc @sidntrivedi012 @starkblaze01

**Comment by sidntrivedi012** [comment]
The `jquery.flot.js` file has this snippet at 135 line :

```Canvas.prototype.resize = function(width, height) {

		if (width <= 0 || height <= 0) {
			throw new Error("Invalid dimensions for plot, width = " + width + ", height = " + height);
		}
```
This is the error that is being shown on the console. 
The line that is causing error in the `capture.js` file : 
```
$W.plot = $.plot($("#graph"),$W.data,flotoptions); 
```
There is something wrong with the plotting of the graph. Looking into it.

**Comment by starkblaze01** [comment]
> The `jquery.flot.js` file has this snippet at 135 line :
> 
> ```
> 
> 		if (width <= 0 || height <= 0) {
> 			throw new Error("Invalid dimensions for plot, width = " + width + ", height = " + height);
> 		}
> ```
> 
> This is the error that is being shown on the console.
> The line that is causing error in the `capture.js` file :
> 
> ```
> $W.plot = $.plot($("#graph"),$W.data,flotoptions); 
> ```
> 
> There is something wrong with the plotting of the graph. Looking into it.

Cool!
Can you verify why width = 0 is passed in the data

**Comment by jywarren** [comment]
So the line where the error is originating (from the last line of the pasted error above) is:

https://github.com/publiclab/spectral-workbench.js/blob/a062b8d45bacfe5e2c202a3a60deca8c60d9c65c/examples/capture/capture.js#L341

Tracking down these lines and getting the permalink and inline code display is really helpful for leaving good notes about your progress, so I recommend it!

**Comment by jywarren** [comment]
Perhaps the thing to do is to try logging out `flotoptions` above this line. Then to as @starkblaze01 suggests, trace back to where `width` is being generated.

**Comment by sidntrivedi012** [comment]
> So the line where the error is originating (from the last line of the pasted error above) is:
> 
> https://github.com/publiclab/spectral-workbench.js/blob/a062b8d45bacfe5e2c202a3a60deca8c60d9c65c/examples/capture/capture.js#L341
> 
> Tracking down these lines and getting the permalink and inline code display is really helpful for leaving good notes about your progress, so I recommend it!

Sure, will keep in mind from the next time.Thanks 🙂

**Comment by sidntrivedi012** [comment]
Whenever I am changing the width here to like a `px` value like 100 px, it is working fine. But in case of `width:100%`, it is giving errors. https://github.com/publiclab/spectral-workbench.js/blob/a062b8d45bacfe5e2c202a3a60deca8c60d9c65c/examples/capture/index.html#L215

**Comment by jywarren** [comment]
Hi, what are the errors exactly? This is odd. But... wait, is flot using a
canvas? this could be the issue. Can you look at how it's working on
spectralworkbench.org for comparison?

**Comment by sidntrivedi012** [comment]
Sure @jywarren. Also, adding to this there's also one more error popping up :
```
Access to XMLHttpRequest at 'file:///capture/recent_calibrations.json?calibration_id=undefined' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, https.
```

**Comment by jywarren** [comment]
OK, that one is where it's trying to fetch recent calibrations from the server. So we'll need to refactor it to accept a constructor parameter on this line:

https://github.com/publiclab/spectral-workbench.js/blob/a062b8d45bacfe5e2c202a3a60deca8c60d9c65c/examples/capture/index.html#L340-L346

And then save it in the initialize method:

https://github.com/publiclab/spectral-workbench.js/blob/a062b8d45bacfe5e2c202a3a60deca8c60d9c65c/examples/capture/capture.js#L34-L37

And then on the line where it's requested, maybe if there is no provided URL, we don't actually send this AJAX request at all:

https://github.com/publiclab/spectral-workbench.js/blob/a062b8d45bacfe5e2c202a3a60deca8c60d9c65c/examples/capture/capture.js#L385-L389

Actually the method is `getRecentCalibrations()` and it's run here:

https://github.com/publiclab/spectral-workbench.js/blob/a062b8d45bacfe5e2c202a3a60deca8c60d9c65c/examples/capture/capture.js#L72-L76

So we could say `if ($W.recentCalibrationsUrl !== false)`

and in initialize, we could have it default to false:
```js
    this.recentCalibrationsUrl = args['recentCalibrationsUrl'] || false
```

**Comment by cryptoclidus** [comment]
I think I got why it's defaulting to 0 when you set width to 100%:

On the html, there are two divs with class="tab pane", with id="settings" and id="capture". the tab pane "settings" is by default active, so capture is inactive (screenshot). So by default, "display:none" for the entire capture tab pane.

This means that any div with no specified width, or min width, (e.g. the elements with width:100% will default to 0 when display:none.). Since the #graph is inside the "capture" tab pane, which is inactive, it's width is default 0.

One possible fix can be to set a min-width of 1 px for the #graph div.

![bug_95_b](https://user-images.githubusercontent.com/30921682/59938343-93fda180-9409-11e9-8c71-93d402c1e082.JPG)

**Comment by cryptoclidus** [comment]
That's also why when you click "Begin Capturing" and the display switches to the "capture" tab pane, the error goes away

**Comment by jywarren** [comment]
Thanks so much for your help! Siddhant how does this look?

**Comment by cryptoclidus** [comment]
Another thought on this: while setting the min-width:1 px is a quick and easy fix, this error is evidence that the program is repeatedly trying to draw the graph, when the graph isn't even being displayed. While it might not really have a substantial effect on performance, is it really necessary to draw the graph when the user is not on the "capture" tab pane? So a more complex fix might be to find all the graph-related and waterfall-related code and put them inside an "if" statement or an event listener that only runs when the "capture" tab pane is active. What do you think?

**Comment by jywarren** [comment]
@sidntrivedi012 can you please check in on this? Thanks @cryptoclidus !!!

**Comment by jywarren** [comment]
Solved in #102!!!

@sidntrivedi012 great work.
